decision_service/dsclientlib/ds_driver.cc

```diff
@@ -0,0 +1,131 @@
+#include "ds_driver.h"
+#include "ds_event.h"
+#include "logger/ds_logger.h"
+
+#include <boost/uuid/uuid_generators.hpp> // generators
+#include <boost/uuid/uuid_io.hpp>         // streaming operators etc.
+#include "error_callback_fn.h"
+
+
+//this macro assumes that success_code equals 0
+#define TRY_OR_RETURN(x) do { \
+  int retval = (x); \
+  if (retval != 0) { \
+    return retval; \
+  } \
+} while (0)
+
+namespace decision_service {
+
+	int check_null_or_empty(const char * arg1, const char * arg2, api_status* status);
+
+	class driver_impl {
+	public:
+    	using error_fn = void(*)(const api_status&,void* user_context);
+
+	private:
+	  	utility::config_collection _configuration;
+    	error_callback_fn _error_cb;
+    	logger _logger;
+
+	public:
+		int ranking_request(const char * uuid, const char * context, ranking_response & response, api_status * status)
+		{
+	    	//clear previous errors if any
+      		if (status != nullptr) status->clear();
```

Comment:
>if (status != nullptr) status->clear(); [](start = 8, length = 39)

introduce a utility function similar to try_set?

Reply:
done

---
In reply to: [186735215](https://github.com/JohnLangford/vowpal_wabbit/pull/1479#discussion_r186735215) [](ancestors = 186735215)